### PR TITLE
Fix signature of adler32(), adler32_combine(), adler32_combine64() and adler32_z() in compat mode.

### PR DESCRIPTION
**Pull request by mtl1979**
* See #700, compat API needs to use "unsigned long" instead of "uint32_t"

### CODE REVIEW

File: zlib.h [hunk 1]
```diff
@@ -1645,7 +1645,7 @@ ZEXTERN void ZEXPORT gzclearerr(gzFile file);
    library.
 */
 
-ZEXTERN uint32_t ZEXPORT adler32(uint32_t adler, const unsigned char *buf, uint32_t len);
+ZEXTERN unsigned long ZEXPORT adler32(unsigned long adler, const unsigned char *buf, unsigned int len);
```
Comment (dankegel) [inline zlib.h:1648]: This seems a bit mysterious.  Won't it be wrong now on 64 bits in the non-compat case?

**Review by dankegel — COMMENTED** [review]
There are a number of other functions that need similar fixes, see output of the script in the bug report.

File: zlib.h (hunk 1 above)
Comment (dankegel) [reply zlib.h:1648]: Seems like we need a type that is ZNG_uLong which expands to 32 bits for ng, and unsized for compat?

Comment (mtl1979) [reply zlib.h:1648]: Non-compat case will not use this header, it uses zlib-ng.h instead.

**Comment by mtl1979** [comment]
@dankegel I did split the changes as several pull requests, one for each source file. See #702 for changes related to ```crc32.c```.

File: zlib.h (hunk 1 above)
Comment (mtl1979) [reply zlib.h:1648]: Using ZNG_uLong isn't enough as some compilers will issue warning or error if function parameters or return type is truncated. Explicit zero-extending is for readability mostly as it makes it clear for human reader and obvious for compiler when inlining -- and static analysis -- that the zero-extending is intended behavior and not oversight.

File: zlib.h [hunk 2]
```diff
@@ -1664,7 +1664,7 @@ ZEXTERN uint32_t ZEXPORT adler32(uint32_t adler, const unsigned char *buf, uint3
      if (adler != original_adler) error();
 */
 
-ZEXTERN uint32_t ZEXPORT adler32_z (uint32_t adler, const unsigned char *buf, size_t len);
+ZEXTERN unsigned long ZEXPORT adler32_z (unsigned long adler, const unsigned char *buf, size_t len);
```
Comment (nmoinvaz) [inline zlib.h]: Prefer no space after function name..

File: adler32.c [hunk 3]
```diff
@@ -116,8 +128,8 @@ static uint32_t adler32_combine_(uint32_t adler1, uint32_t adler2, z_off64_t len
 
 /* ========================================================================= */
 #ifdef ZLIB_COMPAT
-uint32_t ZEXPORT PREFIX(adler32_combine)(uint32_t adler1, uint32_t adler2, z_off_t len2) {
-    return adler32_combine_(adler1, adler2, len2);
+unsigned long ZEXPORT PREFIX(adler32_combine)(unsigned long adler1, unsigned long adler2, z_off_t len2) {
+    return (unsigned long) adler32_combine_((uint32_t) adler1, (uint32_t) adler2, len2);
```
Comment (nmoinvaz) [inline adler32.c]: Prefer no spaces after casts..

File: adler32.c [hunk 4]
```diff
@@ -80,14 +80,26 @@ uint32_t adler32_c(uint32_t adler, const unsigned char *buf, size_t len) {
     return adler | (sum2 << 16);
 }
 
+#ifdef ZLIB_COMPAT
+unsigned long ZEXPORT PREFIX(adler32_z)(unsigned long adler, const unsigned char *buf, size_t len) {
+    return (unsigned long) functable.adler32((uint32_t) adler, buf, len);
```
Comment (nmoinvaz) [inline adler32.c]: No space after casts?

**Review by nmoinvaz — COMMENTED** [review]
Looks pretty good except for some formatting.

File: zlib.h (hunk 2 above)
Comment (mtl1979) [reply zlib.h]: I didn't want to mix style changes with code changes as it is frowned upon...

**Comment by mtl1979** [comment]
Mixing formatting and code changes is frowned upon and that's why I didn't change them at the same time... I tried to be consistent with the style in related crc32.c changes partially already merged in...

**Comment by dankegel** [comment]
That makes abidiff happier, but it still objects to adler32_combine64:

```
  [C]'function uLong adler32_combine64(uLong, uLong, off64_t)' at adler32.c:136:1 has some indirect sub-type changes:
    parameter 1 of type 'typedef uLong' changed:
      typedef name changed from uLong to uint32_t at stdint-uintn.h:26:1
      underlying type 'unsigned long int' changed:
        entity changed from 'unsigned long int' to compatible type 'typedef __uint32_t' at types.h:42:1
          type name changed from 'unsigned long int' to 'unsigned int'
          type size changed from 64 to 32 (in bits)
    parameter 2 of type 'typedef uLong' changed:
      typedef name changed from uLong to uint32_t at stdint-uintn.h:26:1
      underlying type 'unsigned long int' changed:
        entity changed from 'unsigned long int' to compatible type 'typedef __uint32_t' at types.h:42:1
          type name changed from 'unsigned long int' to 'unsigned int'
          type size changed from 64 to 32 (in bits)
```

**Comment by mtl1979** [comment]
@dankegel Updated...

**Comment by dankegel** [comment]
Still present.  Check the last three lines of adler32.c...

**Comment by mtl1979** [comment]
@dankegel Updated once again ;)

**Review by nmoinvaz — APPROVED** [review]

**Review by dankegel — CHANGES_REQUESTED** [review]
I think this code path in zlib.h, which only shows up when you try to compile a program that uses zlib-ng's compat mode from source (and so doesn't show up in the output of abidiff), has the problem, too?
```
--- a/zlib.h
+++ b/zlib.h
@@ -1795,7 +1795,7 @@ ZEXTERN int ZEXPORT gzgetc_(gzFile file);  /* backward compatibility */
      ZEXTERN z_off_t ZEXPORT gzseek64(gzFile, z_off_t, int);
      ZEXTERN z_off_t ZEXPORT gztell64(gzFile);
      ZEXTERN z_off_t ZEXPORT gzoffset64(gzFile);
-     ZEXTERN uint32_t ZEXPORT adler32_combine64(uint32_t, uint32_t, z_off_t);
+     ZEXTERN unsigned long ZEXPORT adler32_combine64(unsigned long, unsigned long, z_off_t);
      ZEXTERN unsigned long ZEXPORT crc32_combine64(unsigned long, unsigned long, z_off_t);
      ZEXTERN void ZEXPORT crc32_combine_gen64(uint32_t *op, z_off64_t);
 #  endif
```
(note also the inconsistent use of z_off64_t there; more on that in #700 ).

Other than that, I don't see any remaining problems.

**Comment by mtl1979** [comment]
@dankegel Updated ;)

**Comment by dankegel** [comment]
LGTM

**Review by nmoinvaz — APPROVED** [review]

**Pull request merged**